src/main/java/core/basesyntax/model/io/DataConverterImpl.java

```
@@ -0,0 +1,69 @@
+package core.basesyntax.model.io;
+
+import core.basesyntax.model.FruitTransaction;
+import java.util.ArrayList;
+import java.util.List;
+
+public class DataConverterImpl implements DataConverter {
+
+    @Override
+    public List<FruitTransaction> convertToTransaction(List<String> lines) {
+        if (lines == null || lines.isEmpty()) {
+            throw new RuntimeException("Brak danych wejściowych do przetworzenia.");
+        }
+
+        List<FruitTransaction> transactions = new ArrayList<>();
+        int lineNumber = 0;
+
+        for (String line : lines) {
+            lineNumber++;
+
+            if (line == null || line.trim().isEmpty()) {
+                throw new RuntimeException("Wiersz " + lineNumber + " jest pusty.");
+            }
+
+            String[] parts = line.split(",");
+
+            if (parts.length != 3) {
+                throw new RuntimeException("Incorrect row format " + lineNumber
+                        + ": 3 fields expected, found " + parts.length);
+            }
+
+            String opCode = parts[0].trim();
+            String fruit = parts[1].trim();
+            String quantityStr = parts[2].trim();
+
+            if (opCode.isEmpty()) {
+                throw new RuntimeException("No operation code in the line " + lineNumber);
+            }
+
+            FruitTransaction.Operation op;
+            try {
+                op = FruitTransaction.Operation.fromCode(opCode);
+            } catch (IllegalArgumentException e) {
+                throw new RuntimeException("Unknown operation code " + opCode
+                        + " in line " + lineNumber);
+            }
+
+            if (fruit.isEmpty()) {
+                throw new RuntimeException("There is no fruit name in the line " + lineNumber);
+            }
+
+            int quantity;
+            try {
+                quantity = Integer.parseInt(quantityStr);
+            } catch (NumberFormatException e) {
```

When parsing quantity you catch NumberFormatException and throw a RuntimeException, but the message is malformed (missing a closing quote) and doesn't include the original CSV line. Per the requirements: "Quantity must be an integer ≥ 0; error on invalid or negative values." and "Do not ignore exceptions; rethrow as RuntimeException with informative messages and original cause." Include the original `line` and pass `e` as the cause (e.g., `new RuntimeException(message, e)`).